Don't destruct element after moving in BlockingReaderWriterQueue's inner_dequeue(..)

kcgen opened the pull request:
Clang's static analyzer warns that the destructor should not be called on the object that's had its content moved out.

![2021-02-12_19-44](https://user-images.githubusercontent.com/1557255/107856165-45212e80-6ddb-11eb-9277-addf2d3483e5.png)

Given this warning was coming from clang-11, I opted to install clang 12 and 13 to see if it was possibly a false-positive (and fixed in newer releases).  This isn't the case and all three versions of clang's static analyzer flag this.

This appears to be a valid issue, given the `element` object should be left in a well-defined state even after its contents have been `std::moved` out.

For example, the `element` object could just as easily have more content moved back into it; or it could allocate new content. Calling the destructor on it closes that door for good though, and doesn't leave `element` in  a workable state.

This logic is also confirmed in [the last paragraph here](https://www.internalpointers.com/post/c-rvalue-references-and-move-semantics-beginners):

---
**A typical move constructor:**

``` c++
Holder(Holder&& other) // <-- rvalue reference in input
{
    m_data = other.m_data;   // (1)
    m_size = other.m_size;
    other.m_data = nullptr;  // (2)
    other.m_size = 0;
}
```

It takes in input an rvalue reference to another Holder object. This is the key part: being an rvalue reference, we can modify it.

So let's steal its data first (1), then set it to null (2). No deep copies here, we have just moved resources around!

It's important to set the rvalue reference data to some valid state (2) to prevent it from being accidentally deleted when the temporary object dies: our Holder destructor calls delete[] m_data, remember?

In general, for reasons that will become more clear in a few paragraphs, it's a good idea to always leave the objects being stolen from in some well-defined state.

---

cameron314 commented:
Yes, I know how move semantics work :-)

In this case, though, `element` is not destroyed because it's moved from, but because its memory will be recycled. The way I've designed the data structure is that a new object is created on `enqueue`, and destroyed on `dequeue`; the only twist is that instead of calling `new` and `delete` (which allocate and free memory in addition to calling the constructor and destructor), I manage my own memory allocation (once, on block creation) and use placement-new and manual destruction instead.

If I wasn't supporting older C++ versions I could call `std::launder` after destruction; I wonder if this would fix the static analysis warning?

kcgen commented:
>  I manage my own memory allocation (once, on block creation) and use placement-new and manual destruction instead.

Absolutely (and Clang has no issue with that).  Ultimately Clang only sees a bonafide reference to T as `element` take issues exclusively with the destruction after `std::move`. So the destructor in this case isn't something you control -- it's provided by T() -- in our case, T is a `std::array<int16_t, 512>`. 

> I could call std::launder after destruction; I wonder if this would fix the static analysis warning?

Unfortunately we're limited to C++14 until more currently-support Linux distros fall off the EOL cliff (perhaps soon enough). (argh; sorry)

cameron314 commented:
> Given `element` is a &T, it's destructor isn't something you control.

I think this is the part I don't understand. Why not? There's no other way to destroy an object instantiated with placement new. `std::array`'s destructor should be called to destroy all the array's elements (a no-op in the case of `std::int16_t`). It can do what it pleases with the memory during the destructor's call, since it hasn't been freed yet.

kcgen commented:
> It can do what it pleases with the memory

`element`will have handed over its (possibly huge) pointers-to-memory into `item`, leaving `element` without any record of or ownership of that memory.  So I would say no; we don't want `element` touching that memory. 

In the purest sense, `std::move` should be considered destructive. 

The donor can either go out of scope (without any further action on your part), or similarly be "donated" into by another donor - essentially re-activating it.

The language authors didn't intend for donors to still be functional after a move.  You don't see code like:

``` c++
    a = std::move(b); // hand over a bit of data
    do_a_bit_more_stuff(b); // change B's processing mode
    clean_up_remaining_stuff(b); // stop a couple of B's threads
    b.~T(); // prevent b from leaking if it goes out of scope
```

That would be disastrous... but I understand C++ does allow for such non-sense with people writing their own move operators.  

The spirit of the move should be atomic and RAII:

```
    a = std::move(b);

    // I never have to touch 'b' again. It won't leak or cause harm or side-effects.
    // All is need to care about is 'a'.

```

The trouble is, people can write terrible C++ classes with broken move semantics. But it shouldn't be your responsibility to manually call the destructor on them post-move.  If their donor objects will leak or cause side-effects post-move, then so be it. 

In our case, I've removed the explicit `element.~T()` cleanup. `element` rotates through tens-of-thousands of rounds of content.  When the program finally terminates, `element` finally goes of out scope and its destructor is naturally call on scope closure. So we have a 1-to-1 relationship between the original "new" to create the original element, and now a single destruction point at scope closure.

dreamer commented:
I *suspect* - in original code, this explicit destructor code is necessary for triggering any side-effects that might be happening via `T` d-tor, right?

ATM *I don't know* if it's necessary in *our* code (I don't think so, but I'm not 100% sure). Perhaps commenting out this d-tor will work (for us).

My opinion is that Clang static analyzer is 100% correct in pointing out that no methods should be called on objects that were `std::move`d away… but I have no idea if this extends to d-tors as well… It feels like riding the edge of what's permissible through C++ standard.

Personally, this code gives me serious pause - I'm afraid it might be completely safe for some types but trigger UB or implementation-defined behaviour for other ones (but it's purely my speculation at this point). If such tricks are necessary, then it perhaps would be better to operate on raw pointers using C constructs…

@kcgen I would prefer to keep such code away from our codebase… but *eh*… if we'll confirm that calling `std::array` destructor this way is safe, then we could keep it and decorate with static asserts to allow usage with good-known types. Considering we're dealing with placement-new here, it might be impossible for Clang static analyzer to detect that explicit d-tor call is safe.

@cameron314 I literally had no idea about the existence of `std::launder` until today. If using it indeed does prevent the static analysis issue - do you plan to decorate the codebase with `#if __cplusplus >= 201703L` or `#if HAVE_LAUNDER`? (In latter case we would need to do add detection and macro-define via buildsystem - it's not a big deal for us - might be a big deal for other users of your library though; former option seems to require jumping over the hoops for MSVC compatibility though…).

kcgen commented:
> then it perhaps would be better to operate on raw pointers using C constructs

It would be easy enough to switch from `std::array` to a C-array, as we're not using any C++'ism's offered by `std::array`. I suppose using a C-scalar type (instead of C++ objects) would eliminate the concern in this bit of code? (the compiler would ignore/elide the `element.~T()` destructor call?).

I can try and see how the static analyzer responds.

kcgen commented:
Let's set aside all the magic of the queue: the internal memory management, reinterpret-casting, and so on - none of that is flagged. Clang's OK with it; and so is LTGM, Coverity, our two dynamic analyzers, along with Valgrind and our manual runtime testing. 

The queue's a template class and the flagged code is bog-standard:

``` c++
T = std::move(ref-to-other T)
ref-to-other.~T()
```

So the burning question is weather the explicit destructor after the move is acceptable code (or an acceptable pattern - given we're deal with template Types.. anything goes. So is this general pattern correct or not?)

If this is true - then everyone everwhere performing a `std::move` on a template type should be able to do the same (explicitly destruct the moved object), just as a rule of thumb.

Do we know the answer to this?

kcgen commented:
> It would be easy enough to switch from std::array to a C-array

I spoke too soon, because I can't simply swap in `int16_t [1024]`'s due to this destructor stuff:

```
./../libs/rwqueue/readerwritercircularbuffer.h:69:53: error: object expression of non-scalar type 'short [1024]' cannot be used in a pseudo-destructor expression
                        reinterpret_cast<T*>(data)[(nextItem + i) & mask].~T();
                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
```

Like you said @dreamer, I would have to switch over to passing around pointers - in which case I can't allocate temporary buffers on the stack because their content really would go out-of-scope (and it doesn't care what's pointing to it). Which then means I need to `new` and `delete` arrays.. and that gets very messy quickly (and ripples across all the audio code and into the soft-limiter).  So yeah.. that's a lot of currently nice-to-read code destroyed just for this.

cameron314 commented:
1. I don't know what exactly the warning is, or why it appears here. I could add `std::launder`, but I don't know if this will help.
2. Removing the call to the destructor will cause undefined behaviour (when reallocating another object on top of the memory of an existing object).
3. It doesn't change the semantics, but from a performance perspective, it's important to note that `std::array<int16_t, 512>` is not movable (it will be fully copied on both enqueue and dequeue). Use `std::vector<int16_t>` or `std::unique_ptr<int16_t[]>` if you want a block of memory that can be moved instantly by swapping pointers.
4. All of the existing code should be 100% safe for all types (for which the code compiles). It's perfectly valid to use `std::array` as the element type (though not very efficient -- see above). Unfortunately my code is not compatible with unwrapped C-style arrays (I'd have to add an overload explicitly for this and I don't see the point with `std::array` being readily available).

## Details:

`std::move` is not destructive. It merely casts to an r-value reference type in order for a move copy constructor/assignment operator to be called, if available. The moved object is still alive and perfectly valid to manipulate, just in an indeterminate state. Like all other C++ objects, it must be destroyed before the memory it occupies can be reclaimed.

With `std::array<int16_t, 512>` as the type, the call to the destructor boils down to zero instructions since `std::array::~array()` only destroys its elements when it is destroyed, but `int16_t` is a POD type that needs no destruction. But the call still needs to be there to avoid undefined behaviour (the destructor must still be called even if it does nothing).

> The language authors didn't intend for donors to still be functional after a move.

Actually yes, they did. That's the whole point around the "valid but indeterminate state" wording. I've personally seen and written code that reuses moved objects... you can't depend on their state, but you can depend on them being valid. This is easy to look up.

Again, here the destructor is *not* called because the object was moved-from. It is called to end the object's lifetime in order to reclaim the memory it is occupying. Two completely independent operations done for independent reasons.

> When the program finally terminates, `element` finally goes of out scope and its destructor is naturally called on scope closure.

Not sure where you get this from. `element` is merely a reference (whose scope ends when the function returns). The referenced object is not destroyed automatically when a reference to it is destroyed. The actual elements in the queue are allocated using placement `new`, and so must be destroyed explicitly (there's no other way). And there's one placement `new` per enqueue (only the memory itself is reused, not the objects).

> If this is true - then everyone performing a `std::move` on a template type should be able to do the same (explicitly destruct the moved object), just as a rule of thumb.

Yes, absolutely. Consider the following code:

    struct Foo { ~Foo() { printf("dtor\n"); } };

    int main() {
        Foo x;
        {
            Foo y = std::move(x);
        }  // y destroyed here, even though it was moved from
    }  // x destroyed here

kcgen commented:
Here's one approach, in a three-step logical transition:

Original ... 
``` c++
	T &element = reinterpret_cast<T*>(data)[i & mask]);
	item = std::move(T);
	element.~T();
```

We can avoid using a reference by `move`ing the cast-memory into a temporary:
``` c++
	T element(std::move(reinterpret_cast<T*>(data)[i & mask]));
	item = std::move(T);
	// The temporary "element" goes out of scope, naturally destructing.
```

We can now eliminate the temporary all together:

``` c++
	item = std::move(reinterpret_cast<T*>(data)[i & mask]));
	// The temporary goes away, and therefore so does the need to destroy any temporary holder.
```

This code works and also passes the analyzer.

cameron314 commented:
@kcgen This transformation does not remove the need to call the destructor. The reference has nothing to do with it. The destructor must be called to destroy the object that is in `data` (remember, `element` is just a reference to the object).

kcgen commented:
> It doesn't change the semantics, but from a performance perspective, it's important to note that std::array<int16_t, 512> is not movable (it will be fully copied on both enqueue and dequeue).

Thanks @cameron314 - I was under the impression it would from this:

![2021-02-13_20-33](https://user-images.githubusercontent.com/1557255/107868799-1a1af700-6e3c-11eb-84aa-07ce9b2530e0.png)

Given std:array can be moved on assignment or construction - does that change things? I'm deliberately using rvalues to facilitate this.  

But if not, switching to vector _is_ just a one-word replacement :-)

cameron314 commented:
Moving a `std::array` will result in all its elements being moved individually, but all N elements of array need to be transferred. There is no pointer to swap (a `std::array` behaves similarly to a struct that wraps a C-style array).

cameron314 commented:
After searching online, it seems this warning from the clang static analyzer appears when methods are called on a moved-from object, despite that being perfectly legal (if uncommon) code. Since the call to the destructor must stay to avoid undefined behaviour, I'm closing this pull request. This is a shortcoming of the analyzer.

kcgen commented:
> After searching online, it seems this warning from the clang static analyzer appears when methods are called on a moved-from object, despite that being perfectly legal

I found a reference here to clang flagging the STL itself as being in violation - https://developercommunity.visualstudio.com/content/problem/549826/stdsort-method-called-on-moved-from-object-of-type.html

It's not much legal we care about - it's correctness and undefined behavior we want to avoid (dereferencing `nullptr` is also legal code-wise). If calling the destructor is infact prevent undefined behavior (especially for poorly defined objects), then great.  I'd rather have a strict guarantee against UB and note that as the reason for accepting this Clang warning.

kcgen commented:
@cameron314 ,

>  I could call std::launder after destruction; I wonder if this would fix the static analysis warning?

I moved our build configuration to C++17,  clang v13 (latest), and added `std::launder` after destruction:

``` c++
	template<typename U>
	void inner_dequeue(U& item)
	{
		std::size_t i = nextItem++;
		T& element = reinterpret_cast<T*>(data)[i & mask];
		item = std::move(element);
		element.~T();
		std::launder(&element); 
		slots->signal();
	}
```

Unfortunately it still warns on the destructor:

``` txt
analyze-build: INFO: ./../libs/rwqueue/readerwritercircularbuffer.h:266:3:
analyze-build: INFO: warning: Method called on moved-from object of type 'std::vector' [cplusplus.Move]
analyze-build: INFO:          element.~T();
analyze-build: INFO:          ^~~~~~~~~~~~
analyze-build: INFO: 1 warning generated.
```

However, destroying via the laundered pointer is permitted, and is happy with this code:

``` c++
	template<typename U>
	void inner_dequeue(U& item)
	{
		std::size_t i = nextItem++;
		T& element = reinterpret_cast<T*>(data)[i & mask];
		item = std::move(element);
		std::launder(&element)->~T();
		slots->signal();
	}
```

You were right :-)  

And just to make sure Clang wasn't simply being misdirected by the pointer, I also tested `(&element)->~T();` but Clang again warned about this.

New PR inbound that uses `std::launder`.

cameron314 commented:
Reading more on `std::launder`, it seems it's more intended for breaking compiler assumptions when reusing the memory of an existing object. It's a bit hairy, but if I understand correctly, I do not need to use it anywhere in my code (since there's no assumptions the compiler can make that would be broken by my straightforward memory reuse). If I were to use it, it would be on the line with the `reinterpret_cast`, and not when calling the destructor.

However, from what I understand it's perfectly legal to use `std::launder` on an ordinary pointer-to-object even when not breaking any typical compiler assumptions, as in your suggested edit.

In this context, then, your proposed call to `std::launder` is legal but unnecessary, yet has four side effects:
1) It prevents the static analyzer from warning about the code
2) It requires a recent C++ standard to be in effect
3) It may prevent the compiler from fully optimizing the code
4) It makes the code significantly more complex to read

While I'm all for suppressing the warning if possible, I'd rather avoid the other three side effects. Since the warning is a false positive in the first place, and there's no "free" workaround, I'm inclined to leave the code as-is (and suppress the warning at the project level via configuration).

kcgen commented:
If it's a false-positive, the yes - I agree with your assessment.

But I'm not sure how to reconcile these points and still believe it's a false-positive:

1) C++11 and 14 documentation doesn't recommend, requiring, or even mention that explicitly calling destructor on a `std::move`d object should be performed to avoid UB. 

    Do you know of such documentation and can you drop a link here?

2) Absent that documentation, I don't see how the clang warning is not legitimate.  The code might be legal: but that's not the goal of a static analyzer.

3) C++17 specifically recommends using `std::launder` to get a pointer to the object located at an address, which is exactly use-case for `reinterpret_cast<T*>(memory-offset)`.  

    `auto p = std::launder(reinterpret_cast<int(*)[10]>(&x[0])); // OK`

    I believe there was some kind of oversight during C++14 and the language authors are having to fix some corner-cases with `launder`.  Otherwise there would be no need for it.

4) When we apply `std::launder` as suggested, the analyzer treats the destructor call as safe.  

If there is UB and we do need to call the destructor, then C++17 appears to be nearest language-junction when `std::launder` allows this. 

But if this is all a false positive, then we should take it up with the clang team.

cameron314 commented:
1. The destructor being called has nothing to do with the object having been moved. It's destroyed because it was created and is no longer needed (this must be done before reusing its memory in a subsequent enqueue operation). Destruction is a normal operation on any object, moved or not, that just happens to be explicit in this case (while normally being implicit). Remember, `std::move` is just a cast that allows the move constructor/assignment operator to be called, which are ordinary methods. By convention, the moved-from object should remain in a valid-but-indeterminate state following the move, but at a minimum it should remain destructible, since it's guaranteed to be destroyed at some point.
2. The goal of the analyzer is to catch smelly code. Use-after-move is rather uncommon, and typically done by accident, and therefore merits a warning. While destroying a moved object is extremely common, doing so explicitly (as done here) is much less so, and unfortunately the analyzer doesn't take that into account as an exception to its rule.
3. Yes, I could add the use of `std::launder` on the lines that `reinterpret_cast`. However, I don't believe it's strictly necessary since there are no assumptions the compiler could make that I would be invalidating. `data` is never used except to get references to the objects that were allocated within it. `std::launder` was added to address shortcomings in earlier standards, which I do not hit. Placement new was around for a long time before `std::launder` and I believe it should work just fine without it in most cases.
4. No, placing `std::launder` *after* `std::move` does not make sense (although is legal).

kcgen commented:
@cameron314  - thanks for taking the time to careful address my concerns. I see and understand your points, and now see the need (and why launder-post-move is hiding this).
Thanks again :+1:

kcgen commented:
Found Clang's logic behind the warning:
https://clang.llvm.org/extra/clang-tidy/checks/bugprone-use-after-move.html

They talk about "re-initialization" of the moved-out-object, if reuse is intended.

It appears that assignment to an empty object counts as re-initializing; likewise the prior object's destruct (or whatever the language calls it in this zombie post-moved-state) must be implicitly called as part.

This code work in C++11/14/17 and passes the analyzer as well:

``` c++
	T& element = reinterpret_cast<T*>(data)[i & mask];
	item = std::move(element);
	element = {};
```

cameron314 commented:
`= {}` is calling the assignment operator (no destructor is called at this point, except possibly on any temporary created with `{}`).

cameron314 commented:
Even with `= {}` and then the explicit destructor call, `= {}` changes the semantics and may result in less optimal code generation (imagine an object which always allocates in its constructor -- this would force an extra allocate/free in that case).

cameron314 commented:
You could still add a dummy method call, though I don't know how far the analyzer traces pointer provenance (if it "sees through" this indirection, and if it doesn't, does the compiler backend itself? We again wouldn't want less optimal code generation):

    item = std::move(element);
    std::forward<T&>(element).~T();

kcgen commented:
> = {} is calling the assignment operator (no destructor is called at this point, except possibly on any temporary created with {}).

Right. We need to trust that the assignment operator does all cleanup needed (and prevent leaks) to accept the inbound assigned content.

That's the same level of trust as we also put in the destructor. 

> may result in less optimal code generation (imagine an object which always allocates in its constructor -- this would force an extra allocate/free in that case).

Right. In both cases, we're calling one extra function post-move.

There could be cases where assignment is cheaper than destruction (I'm not disagreeing with you point, just adding to it).

> You could still add a dummy method call, though I don't know how far the analyzer traces pointer provenance (if it "sees through" this indirection, and if it doesn't, does the compiler backend itself? We again wouldn't want less optimal code generation):
> 
> ```
> item = std::move(element);
> std::forward<T&>(element).~T();
> ```

Gave this a shot, and it still flags it.

cameron314 commented:
My point is that we still have to call the destructor even if we also assign.

cameron314 commented:
Ah well, then I don't see a way to avoid the warning :-(

kcgen commented:
> My point is that we still have to call the destructor even if we also assign.

And my point is that the assignment operator has to guarantee that the object is refreshed and doesn't leak - and is in a workable ("default empty") state.

If we trust the assignment operator, than there is no need to call the destructor (repeatedly; instead assignment will be called repeatedly).

Of course, at the end of the entire program then dequeue goes out of scope and the destructor on those original (reused) object should be called.  I'm not sure if that should be explicit or implicit.

cameron314 commented:
With `element = {}`, there's two objects, both of which need destructing.

kcgen commented:
> With `element = {}`, there's two objects, both of which need destructing.

True; the temporary will be destructed when it goes out of scope (or whatever optimizations the compiler to can do to avoid some/all of that).

kcgen commented:
Ran benchmarks comparing assignment vs. destructor.
Ten rounds, each passing 100M items through the queue.

https://github.com/dosbox-staging/dosbox-staging/pull/862/files

 - Average of 10 run w/ destructor: 4178.3 ms (avg)
 - Average of 10 run w/ assignment operator: 4043.1 ms (avg)

The variation between same-runs is greater than the difference between these aveages - therefore (without even longer running tests and doing a confidence comparison), I see them as the same.

Raw data attached:

[assignment.txt](https://github.com/cameron314/readerwriterqueue/files/5983915/assignment.txt)
[destructor.txt](https://github.com/cameron314/readerwriterqueue/files/5983916/destructor.txt)

cameron314 commented:
Even if the compiler optimizes away the assignment for a specific type, it cannot always optimize it away for all types (for example, any type which allocates memory in its default constructor cannot be optimized out).

kcgen commented:
If a class really needs to (re)allocate just to get back to an "empty state" sufficient even to accept another `std::move` into it, then that cost is going to be borne somewhere and we're just pushing the problem around. 

We all know a well written classes _shouldn't_ allocate on an empty assignment.  Likewise, we imagine costly stuff in the destructor too: for example, a systemd block-device object that spins-up-and-calls `sync` on every mounted file-system within the block device; which it wouldn't do on assignment because it's not shutting down for good.

When I re-read the criteria for [reinitializating](https://clang.llvm.org/extra/clang-tidy/checks/bugprone-use-after-move.html#reinitialization):

![2021-02-15_13-35](https://user-images.githubusercontent.com/1557255/107995796-c15e7200-6f93-11eb-951f-40123bd92dfc.png)

It's pretty simple and clear:

- They could have permitted the destructor to "reinitialize" objects, but they didn't. Why not? 
 
- The destructor doesn't meet the spirit of re-initializing; plain and simple.  It tells the object, _"prepare to be terminated and go away for good"_.  Infact, it very much means the opposite of _initialize_.

- The assignment operator meets the spirit of re-initializing: the object is being assigned fresh content. Especially when you assign a default empty value `{}`; it's as close as you can get to calling the constructor again.

Given the benchmarks results are a wash combined with the undestanding that the assignment operator better meets the spirit of re-initialization, I would lean toward using the default-empty assignment operator over calling the destructor (primarily because it's a better fit for how authors of classes are writing their operators and destructors).

cameron314 commented:
I don't want to call the default constructor and assignment operator, except where necessary, in order to avoid any side effects they may have. I don't need the object to get back to an empty/reinitialized state, I need to destroy it. A new object will be created on enqueue.

cameron314 commented:
Creating an internal object on enqueue, then assigning from it and destroying it on dequeue, is how most similar C++ containers work (for example, `std::queue`).

kcgen commented:
> I don't want to call the default constructor and assignment operator, except where necessary, in order to avoid any side effects they may have. I don't need the object to get back to an empty/reinitialized state, I need to destroy it. A new object will be created on enqueue.

Ahh ok; sorry @cameron314.

 I had thought we were recycling the same objects over and over. I now see this placement in enqueue: `new (reinterpret_cast<T*>(data) + (i & mask)) T(std::forward<U>(item));`, which I trust this is faster or more compatible than std::moving the inbound reference. 

Thanks for sticking with me; I guess we're at the end of the line in terms of exploring options. I thought we had something, but I was on the wrong path assuming we just needed to reinitialize the objects.

cameron314 commented:
No worries, thanks for handling my infinite rebuttals gracefully :-)

Placement new isn't faster (or slower), but it prevents having to new up an array of objects in advance. This also matches the behaviour of other common containers.
